migrator/migrations/m_196_to_m_197_set_poduid_where_null/migration_impl.go

```diff
@@ -0,0 +1,241 @@
+	pgutils.CreateTableFromModel(ctx, database.GormDB, processIndicatorSchema.CreateTableProcessIndicatorsStmt)
+
+	return updateGlobalScope(ctx, database)
+}
+
+// Walk through the process_indicators table and get a map where the key is the id and the value is the
+// PodUid. It doesn't do this for every process indicator, but only for the ones where it could possibly
+// be used to set PodUid in the listening endpoints table.
+func getProcessIndicatorPodUIDMap(ctx context.Context, processIndicatorStore processIndicatorDatastore.Store, processIndicatorIds map[string]bool) (map[string]string, error) {
+	podUIDMap := make(map[string]string)
+
+	err := processIndicatorStore.Walk(ctx,
+		func(processIndicator *storage.ProcessIndicator) error {
+			_, exists := processIndicatorIds[processIndicator.Id]
+			if exists {
+				podUIDMap[processIndicator.Id] = processIndicator.GetPodUid()
+				delete(processIndicatorIds, processIndicator.Id)
+			}
+			return nil
+		})
+
+	processIndicatorIds = make(map[string]bool)
+
+	return podUIDMap, err
+}
+
+// Find out the process indicator ids of plops that could possibly be updated with a poduid from the
+// process_indicators table. This way we don't have to store all id, poduid pairs from the
+// process_indicators table. Go doesn't have sets so use a map where the value is always true instead.
+func getProcessIndicatorIdsOfInterest(ctx context.Context, plopStore plopDatastore.Store) (map[string]bool, error) {
+	processIndicatorIds := make(map[string]bool)
+
+	err := plopStore.Walk(ctx,
+		func(plop *storage.ProcessListeningOnPortStorage) error {
+			if plop.GetPodUid() == "" && plop.GetProcess() == nil {
+				processIndicatorIds[plop.GetProcessIndicatorId()] = true
+			}
+			return nil
+		})
+
+	return processIndicatorIds, err
+}
+
+// Walks through the listening_endpoints table and sets the value of PodUid where it is null using a
+// map obtained from the process_indicators table, where the key is a processindicatorid and the value
+// is a PodUid.
+func setPodUidsUsingProcessIndicators(ctx context.Context, plopStore plopDatastore.Store, podUIDMap map[string]string, batchSize int) error {
+	plops := make([]*storage.ProcessListeningOnPortStorage, batchSize)
+	count := 0
+	err := plopStore.Walk(ctx,
+		func(plop *storage.ProcessListeningOnPortStorage) error {
+			if plop.GetPodUid() == "" {
+				podUID, exists := podUIDMap[plop.GetProcessIndicatorId()]
+				if exists {
+					plop.PodUid = podUID
+					plops[count] = plop
+					count++
+				}
+
+				if count == batchSize {
+					err := plopStore.UpsertMany(ctx, plops)
+					count = 0
+					if err != nil {
+						return err
+					}
+				}
+			}
+
+			return nil
+		})
+
+	if count > 0 {
+		plops = plops[:count]
+		err := plopStore.UpsertMany(ctx, plops)
+		if err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+
+// Sets PodUids by using information from the process_indicators table. It first walks through the
```

Comment:
I initially thought that this strategy of first walking through the listening_endpoints tables and getting a map with processindicatorids that could potentially be used to update PodUids, outlined in this comment was a good idea, but now I am not so sure. At first I thought the map would be small, but I realize now that it could be large, even compared with the number of rows in the process_indicators table.

Reply:
Unfortunately we can't really use an update statement because the data needs to be placed in the `serialized` data as well.